File: app/src/test/java/eu/se_bastiaan/marietje/data/SongsDataManagerTest.java
package eu.se_bastiaan.marietje.data;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Collections;

import eu.se_bastiaan.marietje.data.model.Songs;
import eu.se_bastiaan.marietje.data.remote.SongsService;
import eu.se_bastiaan.marietje.test.common.TestDataFactory;
import rx.Observable;
import rx.observers.TestSubscriber;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class SongsDataManagerTest {

    @Mock
    SongsService mockSongsService;

    private SongsDataManager dataManager;

    @Before
    public void setUp() {
        dataManager = new SongsDataManager(mockSongsService);
    }

    @Test
    public void songsEmitsValues() {
        int currentPage = 0;
        Songs response = stubSongsResponse(currentPage);

        TestSubscriber<Songs> result = new TestSubscriber<>();
        dataManager.songs(currentPage, null).subscribe(result);
        result.assertNoErrors();

        result.assertReceivedOnNext(Collections.singletonList(response));
    }

    @Test
    public void songsCallsApi() {
        int currentPage = 4;
        stubSongsResponse(currentPage);
        dataManager.songs(currentPage, null).subscribe();
        verify(mockSongsService).songs(currentPage, 50, null, null);
    }

    @Test
    public void manageSongsEmitsValues() {
        int currentPage = 1;
        Songs response = stubSongsResponse(currentPage);

        TestSubscriber<Songs> result = new TestSubscriber<>();
        dataManager.manageSongs(currentPage, null, null).subscribe(result);
        result.assertNoErrors();

        result.assertReceivedOnNext(Collections.singletonList(response));
    }

    @Test
    public void manageSongsCallsApi() {
        int currentPage = 2;
        stubSongsResponse(currentPage);
        dataManager.manageSongs(currentPage, null, null).subscribe();
        verify(mockSongsService).manageSongs(currentPage, 50, null, null);
    }

    private Songs stubSongsResponse(int currentPage) {
        Songs response = TestDataFactory.makeSongsResponse(currentPage);
        when(mockSongsService.songs(currentPage, 50, null, null))
                .thenReturn(Observable.just(response));
        when(mockSongsService.manageSongs(currentPage, 50, null, null))
                .thenReturn(Observable.just(response));
        return response;
    }

}

File: app/src/test/java/eu/se_bastiaan/marietje/data/SongsDataManagerTest.kt
package eu.se_bastiaan.marietje.data

import eu.se_bastiaan.marietje.data.model.Songs
import eu.se_bastiaan.marietje.data.remote.SongsService
import eu.se_bastiaan.marietje.test.common.TestDataFactory
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner
import rx.Observable
import rx.observers.TestSubscriber

@RunWith(MockitoJUnitRunner::class)
class SongsDataManagerTest {

    @Mock
    lateinit var mockSongsService: SongsService

    lateinit var dataManager: SongsDataManager

    @Before
    fun setUp() {
        dataManager = SongsDataManager(mockSongsService)
    }

    @Test
    fun songsEmitsValues() {
        val currentPage = 0
        val response = stubSongsResponse(currentPage)

        val result = TestSubscriber<Songs>()
        dataManager.songs(currentPage.toLong(), null).subscribe(result)
        result.assertNoErrors()

        result.assertReceivedOnNext(listOf(response))
    }

    @Test
    fun songsCallsApi() {
        val currentPage = 4
        stubSongsResponse(currentPage)
        dataManager.songs(currentPage.toLong(), null).subscribe()
        verify<SongsService>(mockSongsService).songs(currentPage.toLong(), 50, null, null)
    }

    @Test
    fun manageSongsEmitsValues() {
        val currentPage = 1
        val response = stubSongsResponse(currentPage)

        val result = TestSubscriber<Songs>()
        dataManager.manageSongs(currentPage.toLong(), null, null).subscribe(result)
        result.assertNoErrors()

        result.assertReceivedOnNext(listOf(response))
    }

    @Test
    fun manageSongsCallsApi() {
        val currentPage = 2
        stubSongsResponse(currentPage)
        dataManager.manageSongs(currentPage.toLong(), null, null).subscribe()
        verify<SongsService>(mockSongsService).manageSongs(currentPage.toLong(), 50, null, null)
    }

    private fun stubSongsResponse(currentPage: Int): Songs {
        val response = TestDataFactory.makeSongsResponse(currentPage.toLong())
        `when`(mockSongsService.songs(currentPage.toLong(), 50, null, null))
                .thenReturn(Observable.just(response))
        `when`(mockSongsService.manageSongs(currentPage.toLong(), 50, null, null))
                .thenReturn(Observable.just(response))
        return response
    }

}
